### ISSUE TITLE
Block draggable chip's position is incorrect when dragging over an iframe

### ISSUE THREAD

**Issue by andrewserong** [WordPress/gutenberg#55074]
### Description

The drag chip (the little icon that follows the mouse cursor around) is incorrectly positioned when dragging a block over the site editor canvas, or the post editor canvas when the post editor is iframed (i.e. running TT3 theme with core blocks).

This appears to be to do with the `Draggable` component's positioning when dragging over an iframe in particular.

### Step-by-step reproduction instructions

1. Open the site editor
2. Go to edit a template or the site home
3. Open the block inserter
4. Drag a paragraph block to the editor canvas
5. As you drag over the editor canvas, notice that the drag chip position flicks suddenly to the edge of the browser window

### Screenshots, screen recording, code snippet

https://github.com/WordPress/gutenberg/assets/14988353/619c22e1-3c01-4b12-8760-43de6712ab95

The same issue appears to happen with the Draggable component in Storybook if you use the Docs view that shows all stories. This is a good way to test it because each preview occurs in an iframe, too, it seems. The URL for testing the Draggable component in Storybook is: https://wordpress.github.io/gutenberg/?path=/docs/components-draggable--docs

https://github.com/WordPress/gutenberg/assets/14988353/2be64d83-0456-4e17-b1df-babfeeea5dd7

### Environment info

- Gutenberg trunk on WP 6.3.1 (without GB running, the position is correct in 6.3.1)
- WP 6.4 Beta 1 (without GB running)

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by andrewserong** [comment]
I'm not sure what's caused this one, but from logging out the event in the `Draggable` component's `over` function, it does appear that `e.clientX` / `e.clientY` "reset" when hoving over an iframe so that the coords are relative to the iframe instead of the page 🤔

This is where I was looking: https://github.com/WordPress/gutenberg/blob/f36f4cd74ac4f6722daf4a9dab819e6700b2ce67/packages/components/src/draggable/index.tsx#L188-L203

**Comment by ciampo** [comment]
Mhh, since `Draggable` hasn't changed much recently, I wonder what other change could have caused this. Maybe it's something to do with the editor iframe?

Normally iframes should act as a black hole and absorb all pointer events, but our iframe actually bubbles some of these events ([as I recently learned](https://github.com/WordPress/gutenberg/pull/54164#issuecomment-1740874657)). I can see that #54080 introduced some changes to that functionality, I wonder if it could be related, or at least a good starting point.

**Comment by t-hamano** [comment]
I reported the same issue in #54176, but closed #54176 due to ongoing discussion here.

As @ciampo said, this problem seems to be caused by #54080 (commit hash is [f71f153](https://github.com/WordPress/gutenberg/commit/f71f153642002b6c5b703361d28b7d0a73f49e7b)). Checking out the previous commit ([a247fe8](https://github.com/WordPress/gutenberg/commit/a247fe83dcc29ec316a231f67404be77f61a5ad7)) resolves the issue.

**Comment by andrewserong** [comment]
> I reported the same issue in https://github.com/WordPress/gutenberg/issues/54176, but closed https://github.com/WordPress/gutenberg/issues/54176 due to ongoing discussion here.

Oh, thank you @t-hamano! Apologies for missing your earlier report in that issue, I did a quick skim for issues that already flagged this, but my search was not thorough enough 🙂 

> I can see that https://github.com/WordPress/gutenberg/pull/54080 introduced some changes to that functionality, I wonder if it could be related, or at least a good starting point.

Thank you both for identifying this one as the culprit! That's a big help, I was definitely scratching my head trying to figure out what could've changed 😄

I wonder if we can get the Draggable component to provide an offset if the event is coming from an iframe, so that to a user there's no discernible jump? I'll have a play around.

**Comment by andrewserong** [comment]
Interestingly I found that I can get this issue working properly in the editor again (I think!) if I comment out the check over in `bubbleEvent` before it applies the offset for the event x/y position: https://github.com/WordPress/gutenberg/blob/1ae4d4475c0b1383ed81cd09a9700f1661b78064/packages/block-editor/src/components/iframe/index.js#L41

Without that check, the position seems okay:

https://github.com/WordPress/gutenberg/assets/14988353/37b760e2-f5e7-4dcc-ae58-847188815a88

I do notice there's a bit of a performance hit when dragging over the iframe, possibly due to the cost of the `bubbleEvent` function's `getBoundingClientRect` calls?

I've added a question about the above check in `bubbleEvent` over in https://github.com/WordPress/gutenberg/pull/54080/files#r1348207960, but I'll just ping @youknowriad here, too, for visibility. It'd be good to better understand the intention behind that check, because superficially at least, it seems that if we can fix that check, we might be able to resolve this issue in an elegant way, without having to change how the Draggable component works.

Happy to continue digging on Monday if that doesn't get us there, though!

**Comment by ciampo** [comment]
> the cost of the `bubbleEvent` function's `getBoundingClientRect` calls

I also noticed some jankiness, and calling `getBoundingClientRect` is very likely the reason. Maybe we could "cache" those measurements and update them only when the window resizes (or something similar) ?

**Comment by t-hamano** [comment]
I found that I could solve the problem by changing the following:

```diff
diff --git a/packages/block-editor/src/components/iframe/index.js b/packages/block-editor/src/components/iframe/index.js
index 9a6371dcaf..badbbfd2b9 100644
--- a/packages/block-editor/src/components/iframe/index.js
+++ b/packages/block-editor/src/components/iframe/index.js
@@ -38,7 +38,7 @@ function bubbleEvent( event, Constructor, frame ) {
                init[ key ] = event[ key ];
        }
 
-       if ( event instanceof frame.ownerDocument.defaultView.MouseEvent ) {
+       if ( event instanceof frame.contentDocument.defaultView.MouseEvent ) {
                const rect = frame.getBoundingClientRect();
                init.clientX += rect.left;
                init.clientY += rect.top;
```

Before being refactored in #54080, it [appears](https://github.com/WordPress/gutenberg/pull/54080/files#diff-ab17abd38e687edd56573694ae3017e105e96b319571c190e9d4705c1872bc9cL59) to be checking if the current event is an instance of the iframe document _itself_. On the other hand, the refactored code seems to check if the event is an instance of the iframe's _parent_ document.

So should we check the current document (`currentDocument`) instead of the parent document (`ownerDocument`)?

**Comment by andrewserong** [comment]
@t-hamano we both found the same thing at the same time! I have a PR up over in #55150 😄

Note: it doesn't address the issue of Storybook in the Docs view, since Storybook doesn't receive the offset logic in our Iframe component. I reckon we could leave that issue aside for now, as fixing that would likely involve moving the offset logic to the `Draggable` component, and it'd probably be better for us to go with a simple fix in the short-term for WP 6.4.
